odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+        If ``other`` is an operator, since a functional is also an operator
+        this corresponds to operator composition:
+
+            ``op * func <==> (x --> op(func(x))``
+
+        If ``other`` is a scalar, this corresponds to left multiplication of
+        scalars with functionals:
+
+            ``scalar * func <==> (x --> scalar * func(x))``
+
+        If ``other`` is a vector,  since a functional is also an operator this
+        corresponds to left multiplication of vectors with operators:
+
+            ``vector * func <==> (x --> vector * func(x))``
+
+        Note that left and right multiplications are generally
+        different.
+
+        Parameters
+        ----------
+        other : `Operator` or `LinearSpaceVector` or scalar
+            `Operator`:
+            The `Operator.domain` of ``other`` must match this functional's
+            `Functional.range`.
+
+            `LinearSpaceVector`:
+            ``other`` must be an element of this functionals's
+            `Functional.range`.
+
+            scalar:
+            The `Operator.domain` of this operator must be a
+            `LinearSpace` and ``other`` must be an
+            element of the ``field`` of this operator's
+            `Operator.domain`.
+
+        Returns
+        -------
+        rmul : `Functional` or `Operator`
+            Multiplication result
+
+            If ``other`` is an `Operator`, ``rmul`` is a `OperatorComp`.
+
+            If ``other`` is a scalar, ``rmul`` is a
+            `FunctionalLeftScalarMult`.
+
+            If ``other`` is a vector, ``rmul`` is a
+            `OperatorLeftVectorMult`.
+        """
+        if other in self.domain.field:
+            if other == 0:
+                from odl.solvers.functional.default_functionals import (
+                    ZeroFunctional)
+                return ZeroFunctional(self.domain)
+            else:
+                return FunctionalLeftScalarMult(self, other)
+        else:
+            return super().__rmul__(other)
+
+    def __add__(self, other):
+        """Return ``self + other``.
+
+        Parameters
+        ----------
+        other : `Functional` or scalar
+            `Functional`:
+            The `Functional.domain` and `Functional.range` of ``other``
+            must match this functional's  `Functional.domain` and
+            `Functional.range`.
+
+            scalar:
+            The scalar needs to be in this functional's `Functional.range`.
+
+        Returns
+        -------
+        add : `Functional`
+            Addition result
+
+            If ``other`` is in ``Functional.range``, ``add`` is a
+            `FunctionalScalarSum`.
+
+            If ``other`` is a `Functional`, ``add`` is a `FunctionalSum`.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return NotImplemented
```

Comment:
```OperatorSum``` chokes on this: 

```python
=================================== FAILURES ===================================
_________________________ test_functional_plus_scalar __________________________

    def test_functional_plus_scalar():
        """Test for sum of functioanl and scalar."""
        space = odl.uniform_discr(0, 1, 10)
    
        func = odl.solvers.L2NormSquared(space)
        scalar = np.random.randn()
    
        # Test for scalar not in the field (field of unifor_discr is RealNumbers)
        complex_scalar = np.random.randn() + np.random.randn() * 1j
        with pytest.raises(TypeError):
>           func + complex_scalar

test/solvers/functional/functional_test.py:228: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
odl/solvers/functional/functional.py:354: in __add__
    return super().__add__(other)
odl/operator/operator.py:732: in __add__
    return OperatorSum(self, other)
[...]
>       if left.range != right.range:
E       AttributeError: 'complex' object has no attribute 'range'
```
`Operator.__add__` just passes through almost all cases to `OperatorSum`:
```python
    def __add__(self, other):
        """[...]
        """
        from odl.operator.default_ops import ConstantOperator

        if other in self.range:
            return self + ConstantOperator(other, self.domain, self.range)
        elif other in self.range.field:
            constant_vector = other * self.range.one()
            return self + ConstantOperator(constant_vector, self.domain,
                                           self.range)
        else:
            return OperatorSum(self, other)
```

 and `OperatorSum` cannot handle it:
```python
def __init__(self, left, right, tmp_ran=None, tmp_dom=None):
        """[...]
        """
        if left.range != right.range:
            raise TypeError('operator ranges {!r} and {!r} do not match'
                            ''.format(left.range, right.range))
```

 It is related to #563 and the corresponding pull request.